### ISSUE TITLE
Support for deriving `OneofObject` and `Union` +  `OneofObject` and `Interface` simultaneously

### ISSUE THREAD

**Issue by bvanneerven** [async-graphql/async-graphql#1005]
## Description of the feature
Since the `@oneOf` directive only applies to input now (also see #915), it would be nice if other forms of output polymorphism could be supported using the same enum we use for `OneofObject`. This could be achieved by allowing to simultaneously derive `OneofObject` and `Union` or `OneofObject` and `Interface`.

This seems to be possible by (re)using `#[graphql(input_name = "...")]` (see the example below). Right now, derivation at the same time is not possible because of a runtime error: 

> thread 'main' panicked at 'Register `Ab` as `Union`, but it is already registered as `InputObject`

When using `OneofObject` and `Interface` simultaneously we get a compile time error:

> error: Unknown field: `field`

### `OneofObject` and `Union`
```rs
use async_graphql::*;

#[derive(SimpleObject, InputObject)]
#[graphql(input_name = "AInput")]
struct A {
    a: String,
}

#[derive(SimpleObject, InputObject)]
#[graphql(input_name = "BInput")]
struct B {
    b: String,
}

#[derive(Union, OneofObject)]
#[graphql(name = "AB")]
#[graphql(input_name = "ABInput")]
enum AB {
    A(A),
    B(B),
}
```

```gql
directive @oneOf on INPUT_OBJECT

type A { 
    a: String! 
}
input AInput { 
    a: String! 
}

type B { 
    b: String! 
}
input BInput { 
    b: String! 
}

union AB = A | B
input ABInput @oneOf {
    a: AInput
    b: BInput
}
```

### `OneofObject` and `Interface`
```rs
use async_graphql::*;

#[derive(SimpleObject, InputObject)]
#[graphql(input_name = "AInput")]
struct A {
    name: String,
    a: String
}

#[derive(SimpleObject, InputObject)]
#[graphql(input_name = "BInput")]
struct B {
    name: String,
    b: String
}

#[derive(Interface, OneofObject)]
#[graphql(name = "AB")]
#[graphql(input_name = "ABInput")]
#[graphql(field(name = "name", type = "String"))]
enum AB {
    A(A),
    B(B),
}
```
```gql
directive @oneOf on INPUT_OBJECT

type A implements AB { 
    name: String!
    a: String! 
}
input AInput { 
    name: String!
    a: String! 
}

type B implements AB { 
    name: String!
    b: String! 
}
input BInput { 
    name: String!
    b: String! 
}

interface AB {
    name: String!
}
input ABInput @oneOf {
    a: AInput
    b: BInput
}
```

**Comment by sunli829** [comment]
Using multiple procedural macros at the same time and avoiding conflicts is often cumbersome and requires a lot of compatibility work. 😂

**Comment by bvanneerven** [comment]
Partially solved in #1116
